firebase-firestore/src/main/java/com/google/firebase/firestore/FirebaseFirestore.java

```diff
@@ -161,20 +161,34 @@ public FirebaseFirestoreSettings getFirestoreSettings() {
    */
   @PublicApi
   public void setFirestoreSettings(@NonNull FirebaseFirestoreSettings settings) {
-    checkNotNull(settings, "Provided settings must not be null.");
-    // As a special exception, don't throw if the same settings are passed repeatedly. This
-    // should make it simpler to get a Firestore instance in an activity.
-    if (client != null && !this.settings.equals(settings)) {
-      throw new IllegalStateException(
-          "FirebaseFirestore has already been started and its settings can no longer be changed. "
-              + "You can only call setFirestoreSettings() before calling any other methods on a "
-              + "FirebaseFirestore object.");
+    synchronized (databaseId) {
+      checkNotNull(settings, "Provided settings must not be null.");
+      // As a special exception, don't throw if the same settings are passed repeatedly. This
+      // should make it simpler to get a Firestore instance in an activity.
+      if (client != null && !this.settings.equals(settings)) {
+        throw new IllegalStateException(
+            "FirebaseFirestore has already been started and its settings can no longer be changed. "
+                + "You can only call setFirestoreSettings() before calling any other methods on a "
+                + "FirebaseFirestore object.");
+      }
+      this.settings = settings;
     }
-    this.settings = settings;
+  }
+
+  /**
+   * If this returns true, then Firestore is configured. This is not synchronized. So it may return
+   * false when either Firestore is not configured or is being configured.
```

Comment:
As written, this isn't quite true. I think you need to mark `client` as `volatile`. Checkout the java section of this: https://en.wikipedia.org/wiki/Double-checked_locking for more details.

(Summary: java allocates memory for client (which makes it non-null) before initializing the memory. So you could end up with client != null, but not valid either.)

Reply:
thank you for pointed out. A boolean member (of primary type) is added.

Reply:
Sorry for the slow turn-around but this part of the change isn't ideal.

The whole point of the `ensureClientConfigured` call is to do what's needed (in a single call) to ensure that subsequent use of the client will work. The idea is to reduce the boilerplate and chance for error at all the call sites that need to deal with deferred initialization.

As a result I think that requiring callers to check `isConfigured` first goes against that mission. We should move this check into that method (but outside the synchronized block).

Secondarily, because this exists to guide the implementation of already private methods this opens up an api surface area that could be misunderstood. For example, it's not unreasonable for someone to check `isConfigured()` and then unintentionally violate this contract by just using the `client` directly while forgetting to call `ensureClientConfigured`. I'm not saying that such a usage is correct or likely but rather we should design even internal APIs against misuse. In particular we should attempt to avoid multi-step conditional contracts like these and instead try to come up with simple ways to convey the correct way to use it.

Finally, changing something like this should be subject to heightened scrutiny because it needs to be replicated across all the clients. If there's a way to avoid changing an internal contract (and update all callers we should avoid doing so).